### DISCUSSION TITLE
Issue with invoking server action within a server component for deleting a cookie

### DISCUSSION THREAD

**Discussion by riculum** [vercel/next.js#52096] [Help] [answered]
### Summary

I have been trying to implement a Sign Out page in my Next.js project, where I need to delete a cookie. However, I keep encountering an error message: "Error: Cookies can only be modified in a Server Action or Route Handler." I have already set up the `serverActions` in my `next.config.js` file, but the error persists.

```
import {cookies} from "next/headers";

async function deleteCookie() {
    'use server'
    cookies().delete('accessToken')
}

export default async function SignOut() {
    await deleteCookie()

    return (
        <>
            <h1>SignOut</h1>
        </>
    )
}
```

Nextjs version: 13.4.8-canary.14

**Comment by rahXephonz** [comment]
I think better safe clear cookie in client side instead of server side

instead of that i have another example
using js-cookie

jsCookie.remove(‘access-token’) this fn works if you work an interactive components like click button signout or etc

**Reply by riculum** [reply]
This may work for the logout area, but for the login the cookie would have to be set as httponly attribute for security reasons, which does not work on the client side.

**Reply by riculum** [reply]
I just tried also with js-cookie to delete the accessToken. Since this is httpyOnly as just mentioned, it cannot be deleted.

**Reply by rahXephonz** [reply]
can you give me link example for repro?
maybe i can help you for this case.

**Comment by icyJoseph** [comment]
I reckon the problem is that you want to do this, __as you render__, and not on a user activated handle (or programmatically activated by you).

So you'd need to place this server action, as the `action` prop on a form,  or a use client component submit button, and do it from there, so that it happens from the client, and not as you render on the server.

What's the user flow you have? Does the user need to click a button on this SignOut Page to log out? or is this page just a placeholder?

**Reply by riculum** [reply]
The page is a kind of placeholder (without any user interaction). When the user navigate to /logout, all cookies should be deleted (including all httpOnly, which is only possible server-side) and the user should be redirected to the login page.

**Reply by icyJoseph** [reply, marked as answer]
Then you are looking at a handful of options.

- Least recommended: When the user arrives to this logout route, inside an effect call the server actions, or maybe event set a timer
- In a similar fashion, you could call an endpoint to delete the cookie (API route)
- Most favourable: Use middleware, as the user transitions to logout the middleware should delete the cookie, https://nextjs.org/docs/app/building-your-application/routing/middleware#using-cookies

**Reply by icyJoseph** [reply]
I think you'd need to do something like:

```ts
import {NextRequest, NextResponse} from 'next/server'

export async function middleware(req: NextRequest) {
    if (req.url.includes("/signout")) {
	const response = NextResponse.redirect(new URL('/signin', request.url))

        response.cookies.delete("accessToken")
        response.cookies.delete("refreshToken")

	return response;
    }

    return NextResponse.next()
}

export const config = {
    matcher: ['/signout', '/signin'],
}

```

**Comment by sonyarianto** [comment]
@icyJoseph what do you think? better using middleware or using route handler (API route)? Route handler seems straight forward
